### PR TITLE
JS: Add model of pg-promise

### PR DESCRIPTION
**Pull request by asgerf**
Adds a model of [pg-promise](https://www.npmjs.com/package/pg-promise), a PostgreSQL connector library.

Evaluations: (internal link)
- [On 10 projects using pg-promise](https://github.com/dsp-testing/asgerf-dca/blob/run/js/pg-promise3/reports) shows 231 new sinks and 30 new alerts.
- [Default slugs](https://github.com/dsp-testing/asgerf-dca/tree/run/js/pg-promise2/reports) from a slightly older version of the PR shows acceptable performance.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 1]
```diff
@@ -137,17 +149,126 @@ private module Postgres {
...
+    private DataFlow::Node getADirectQueryArgument() {
+      result = getArgument(0)
+      or
+      result = getOptionArgument(0, "text")
+    }
+
+    /**
+     * Gets an interpolation parameter whose value is interpreted literally, or is not escaped appropriately for its context.
+     *
+     * For example, the following are raw placeholders: $1:raw, $1^, ${prop}:raw, $(prop)^
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll:215]: TODO: Check if existing DB models have a `raw` interpolation mechanism.

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 2]
```diff
@@ -96,28 +96,40 @@ private module MySql {
  * Provides classes modelling the `pg` package.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll:96]:
```suggestion
 * Provides classes modelling the PostgreSQL packages, such as `pg` and `pg-promise`.
```

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 3]
```diff
@@ -137,17 +149,126 @@ private module Postgres {
...
+
+  /** Gets a connection created from a `pg-promise` database. */
+  API::Node pgpConnection() {
+    result = pgpDatabase().getMember("connect").getReturn().getPromised()
+  }
+
+  /** Gets a `pg-promise` task object. */
+  API::Node pgpTask() {
+    exists(API::Node taskMethod |
+      taskMethod = pgpObject().getMember(["task", "taskIf", "tx", "txIf"])
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll:182]: Can we add a test for one of these?

**Review by esbena — COMMENTED** [review]
LGTM. Impressive numbers!

#### Subjective module structuring:

What do you think of splittling the `Postgres` module into two internal modules `Postgres::pg` and `Postgres::pg-promise` to avoid the explicit mention of `pg-promise` in all of the docstrings?

#### Types

`pg-promise` has [types](https://github.com/vitaly-t/pg-promise/blob/master/typescript/pg-promise.d.ts), can we add the `ofType` cases to this PR, like we do for our related `mongodb`/`mongoose` models?

https://github.com/github/codeql/blob/b381f4826c0f34998a378cb073f3ac1813afc3bf/javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll#L224

**Comment by asgerf** [comment]
> What do you think of splittling the Postgres module into two internal modules Postgres::pg and Postgres::pg-promise to avoid the explicit mention of pg-promise in all of the docstrings?

I don't think it's a huge problem to mention `pg-promise` in the docstring. It's also easier to read than having to scroll upwards to see what module you're inside.

> pg-promise has types, can we add the ofType cases to this PR, like we do for our related mongodb/mongoose models?

👍

File: javascript/ql/test/query-tests/Security/CWE-089/untyped/pg-promise-types.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { IDatabase } from "pg-promise";
```
Comment (esbena) [inline javascript/ql/test/query-tests/Security/CWE-089/untyped/pg-promise-types.ts:1]: nit: this file belongs to the `typed` directory

Comment (asgerf) [reply javascript/ql/test/query-tests/Security/CWE-089/untyped/pg-promise-types.ts:1]: The `typed` directory is for tests that rely on type information provided by the TypeScript compiler, which in turn requires a `.d.ts` shim.

As of https://github.com/github/codeql/pull/4401, however, we don't need that in our tests anymore, which is why the test works in this directory.

**Review by esbena — DISMISSED** [review]

**Review by esbena — APPROVED** [review]

**Pull request merged**